README
   =====================================================================

                    PyRT: Python Routeing Toolkit

   Copyright (C) 2001 Richard Mortier <mort@sprintlabs.com>, Sprint ATL

   =====================================================================

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
   USA

   =====================================================================

1. Purpose
   -------

   The purpose of this software is to enable routeing information in a
   network (ie. Sprintlink) to be collected.  This package currently
   supports BGPv4 and ISIS and will dump MRTD format files.  It also
   supports parsing of MRTD TABLE_DUMP files (as available from, eg.,
   RouteViews and RIPE/RIS).  A number of utilities for manipulating
   these dumps are also provided.

   PyRT is distributed under the terms of the GNU General Public
   License version 2.  A copy is contained within this distribution in
   the file COPYING.

   =====================================================================

2. Overview
   --------

   All code is written in Python v2(.1.1) and was developed under
   Redhat 7.1.  The code is split modules which may be imported and
   are described in detail below.

   A module consists of 4 parts:

   ** imports, global variables, constants, etc;
   ** module exported functions available to other programs;
   ** a wrapper class to abstract communication details and lower
      layers;
   ** an entry point ('if __name__ == "__main__": <code>').

   Protocol modules contain all 4 parts; other modules may not.

   It was tested against Zebra 0.91a and a CISCO 7xxx running IOS 12.1.

   =====================================================================

3. Code structure

     -------------------------------------------------------------------

3.0. mutils.py

     This is a generic utility module.  It provides some handy
     functions for manipulating IP addresses/prefixes, and for
     printing packed strings in hexadecimal and binary.  The naming
     convention is:

     ** mask2plen/addrmask2str:
            manipulate and print address/mask prefixes (as opposed to
            prefix/length prefixes).
            
     ** str2id/id2str/pfx2str/str2pfx: 
            convert between string and long representations of IP
            addresses/prefixes.

     ** isid2id:
            given an ISIS/CLNP address, convert to the numeric
            representation of the corresponding IP address (assuming
            the standard 11.22.33.44.55.66 -> 112.233.445.566
            mapping).

     ** int2xxx: 
            given an integer, return a string representation in base
            xxx.

     ** str2xxx: 
            given a string (ie. packed binary data a la the 'struct'
            module), return a string representation in base xxx.

     ** prtxxx: 
            print the string representation, taking care of wrapping
            and prepending a prefix to each line.

     -------------------------------------------------------------------

3.1. Protocol modules

     These provide support for particular routeing protocols; current
     ISIS and BGPv4 are supported.  Each exports functions for parsing
     messages associated with that protocol, in addition to a class
     which enables communication with other entities supporting that
     protocol for data collection purposes, and an entry point.  In
     general, the classes do not implement protocol state machines
     completely, but do the minimum necessary to start receiving
     routeing information.  Similarly, error handling is (just about)
     non-existant.  No class will advertise any routeing information.

     All protocol modules essentially work in the same manner:

     new protocol()
     initialize protocol
     while 1:
        protocol.parse()

     where the protocol constructor sets up the object, protocol
     initialization involves doing whatever needs to be done to talk
     to an entity supporting that protocol, and protocol.parse() uses
     the protocol class to read the next message in appropriately, and
     then dump and/or print it accordingly.  All listeners support 4
     verbosity options:

     ** quiet (-q): 
            output nothing.

     ** default: 
            output human readable parsing of protocol messages.

     ** verbose (-v): 
            in addition to default, output significant chunks of
            messages in binary/hex.

     ** very verbose (-V): 
            in addition to above, output results/parameters of
            recvMsg(), parseMsg(), sendMsg() (mostly for debugging
            purposes).

       -----------------------------------------------------------------

3.1.0. MRTD (mrtd.py)

       This module implements some generic logging functionality.  It
       provides support for reading and writing MRTD format log files
       in 3 formats:

       ** PROTOCOL_BGP: 
              default for the BGP module; slightly extended to allow
              messages other than UPDATEs to be dumped.

       ** PROTOCOL_BGP4MP: 
              alternative BGP dump format used by Zebra. Contains
              hacks to work around problems in Zebra dumps (v0.89 and
              0.91a).

       ** PROTOCOL_ISIS: 
              non-standard homegrown format for dumping ISIS messages.

       It provides a class to wrap up dealing with reading and writing
       MRTD dumps.  This handles details such as dump file rotation
       based on size limits.  It also attempts to be mildly efficient
       by buffering reads, whilst ensuring that there is always a
       complete MRTD message available to be parsed.  Writes are
       unbuffered at the present time.

       It also supports (with the aid of the BGP module) parsing of
       MRTD TABLE_DUMP files.

       -----------------------------------------------------------------

3.1.1. BGP v4 (bgp.py)

       This module implements the BGP listener.  It provides extensive
       parsing support, but implements a highly restricted subset of
       the state machine.  Essentially:

       initialize
       socket connect to remote end
       send OPEN
       receive OPEN
       send KEEPALIVE ## remote end should be ESTABLISHED at this point
       while 1:
           parse message

       Although all messages are parsed, NOTIFYs are not obeyed, and
       in normal operation we expect to receive only UPDATEs, and
       transmit nothing, after the connection becomes ESTABLISHED.  If
       an error occurs at any point, the script will either exit
       gracelessly or sit there waiting forever, and need to be
       restarted.

       A note on Bgp.recvMsg(): this tries to be (slightly) cunning.
       Since BGP PDUs are transported over TCP (a byte-stream with no
       PDU boundary information), we have to ensure that we read
       enough data off the socket to get a complete BGP PDU.  This is
       done by repeatedly reading into a per-instance buffer until we
       are sure that there is enough data in the buffer that the
       complete PDU can be recovered.  This has managed to be the most
       bug-ridden part of this module so far, so watch out...

       Eg.

       : $; ./bgp.py -p 10.64.233.1 -a 200 --local 10.64.233.42 -m \
                     -f bgp-dump -z $((1024*1024*5))

       -----------------------------------------------------------------

3.1.2. ISIS (isis.py)

       The module implements the ISIS listener.  It provides parsing
       support for most ISIS messages, and implements as little of the
       state machine as I could (damn those ISO protocol writers for
       being too good :-)).  None of the clever stuff about rate
       limiting transmissions, jittering timers, and so on is
       performed (we should never be transmitting so many packets that
       anyone would care).  Essentially it forms a single adjacency
       with the router to which the machine is physically connected.
       This adjacency is maintained as mandated using HELLOs
       (currently only LAN HELLOs are supported).  It then listens for
       and dumps all packets received and transmitted.  Basically:

       while 1:
           select on (message-to-read, timeout-expired)

           if message-to-read:
               Isis.parseMsg()

           else: ## we need to transmit one or more packets 
               for-each adjacency:
                   if required:
                       transmit ISIS HELLO
                   update timeout

       Since ISIS is transported directly in zero padded 802.2 frames,
       message boundaries are frame boundaries so we don't need any
       buffering cunningness a la BGP.  However, since we have 802.2
       frames, messages are retrieved via a raw socket (PF_PACKET,
       SOCK_RAW), the reason this listener requires Python v2(.1.1).
       See the comment at the start of the module for an explanation
       of the MAC/LLC encapsulation used.

       Eg.

       : $; ./isis.py -d -a 49.00.01 -z $((1024*1024*5)) -f isis-dump

       -----------------------------------------------------------------

3.1.3. OSPF (ospf.py)

       This module implements the OSPF listener, blah, as for the ISIS
       module :)
     
     -------------------------------------------------------------------

3.2. Utilities

     In addition to the example parse.py, a small number of utilities
     are also provided in the scripts subdirectory.  In no particular
     order:

       -----------------------------------------------------------------

3.2.0. parse.py

       Dummy script that demonstrates basic use of the MRTD/protocol
       libraries and dictionary return formats.

       -----------------------------------------------------------------

3.2.1. splice.py

       This takes a number of MRTD files on the command line, and
       splices their messages together in time order (breaking ties by
       the order they appeared on the command line).  It can output to
       files of a given size, using the same rotation algorithm as the
       MRTD module.

       -----------------------------------------------------------------

3.2.2. clean.py

       This cleans a trace file, ensuring that a valid rv is returned
       for each message (at least, an rv with non-None "T" field).

   =====================================================================

4. References

   BGP: "BGP4: Inter-domain routing in the Internet" John W. Stewart III
   
        RFC 1771 (base) 
        RFC 1965 (confederations) 
        RFC 1966 (route reflectors)
        RFC 1997 (communities)
        RFC 2283, RFC 2858 (multi-protocol extensions)
        RFC 2842 (capabilities advertisement)
        RFC 2918 (route refresh)

        draft-ietf-idr-cease-subcode-00.txt (cease notification subcodes)
        draft-ietf-idr-bgp-ext-communities-01.txt (extended communities)

   MRTD: http://www.mrtd.net/
         Zebra documentation

   ISIS: RFC 1142 (OSI ISO ISIS)
         RFC 1195 (IP extensions)
         RFC 2763 (Dynamic hostname TLV)
         http://www.gated.org/ 
             GateD public release v3.5.11 (the last to support ISIS,
             and then only on NetBSD)
         http://www.rware.demon.co.uk/isis.htm
             magic numbers, MAC/LLC headers

         http://www.cisco.com/warp/public/97/tlvs_5739.html

   Zebra: http://www.zebra.org/

   =====================================================================

5. To do
   
   ** need to _completely_ separate the parsing from pretty printing
   ** ext timestamp support is ugly -- should push into common header
   ** bgp: support KEEPALIVEs for robustness

   ** write buffering
   ** bgp: support multiple peering sessions
   ** isis: fix the finding of the IP address (ie. make -i obsolete)

========================================================================

$Id: README,v 1.15 2002/01/23 23:17:17 mort Exp $

========================================================================


ChangeLog
2004-06-16  Richard Mortier  <mort@cantab.net>

	* (ospf.py) now have OSPF listener too; incomplete (hence v2.9)
	  until I can test against a router, but parsing code seems to be
	  done.  
	
	* (mrtd.py, mutils.py) bug fixes (int2bin() partic.), also more
	  robust to existence/non-existence of modules and features
	
	* new release: 2.9

2002-05-05  Richard Mortier  <mort@ieee.org>

	* (bgp.py) extended length attributes actually didn't work; now
	  they do (tickled by parsing recent RouteViews data where
	  MP_REACH_NLRI and MP_UNREACH_NLRI fields are to be found)

	* (bgp.py) also fixed situations when NLRI field was prematurely
	  truncated (I believe this relates to a Zebra bug in
	  bgpd/bgp_dump.c where the allocated buffer doesn't allow space
	  for the type specific header in addition to max. packet size and
	  common header; gets tickled by UPDATES which are close to max
	  packet size) 

	* new release: 2.5

2002-02-25  Richard Mortier  <mort@sprintlabs.com>

	* finished table-dump.py, few bug fixes

        * failure detection scripts + matlab; weight change scripts

        * new release: 2.4

2002-02-05  Richard Mortier  <mort@sprintlabs.com>

        * new release: 2.3

	* (table-dump.py) added table-dump.py enabling creation of RIB
	  dumps from (single) UPDATE streams
	
        * (splice.py) ditto

        * (mrtd.py) realised that timestamping is confused -- MRT
          packet timestamps are seconds-since-UTC, filenames are
          localtime; henceforth, all shall be based on
          seconds-since-UTC.
	
	* ChangeLog begins!  released version is 2.2

   =====================================================================

			 PyRT: Python Routeing Toolkit

				   ChangeLog
                                   
	      Copyright (C) 2001 Sprint ATL.  All Rights Reserved.

   =====================================================================


table-dump.py
#! /usr/bin/env python2.5

##     PyRT: Python Routeing Toolkit

##     Manipulates MRTd dumps, splicing one or more dumps together in
##     time order, and outputting the result given start/end times,
##     and given a certain 'chunk' size.

##     Copyright (C) 2001 Richard Mortier <mort@sprintlabs.com>, Sprint ATL

##     This program is free software; you can redistribute it and/or
##     modify it under the terms of the GNU General Public License as
##     published by the Free Software Foundation; either version 2 of the
##     License, or (at your option) any later version.

##     This program is distributed in the hope that it will be useful,
##     but WITHOUT ANY WARRANTY; without even the implied warranty of
##     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
##     General Public License for more details.

##     You should have received a copy of the GNU General Public License
##     along with this program; if not, write to the Free Software
##     Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
##     02111-1307 USA

import time, getopt, sys, string, os, pprint
import mrtd, bgp
from mutils import *

VIEW_NO      = 0x00
STATUS       = 0x01
MINS_TO_SECS = 60

################################################################################

def processEntry(rv):

    def mkStr(attr, aval):
    
        flags = 0
        if aval["FLAGS"]['optional']:
            flags = flags | (1<<7)
        if aval["FLAGS"]['transitive']:
            flags = flags | (1<<6)
        if aval["FLAGS"]['partial']:
            flags = flags | (1<<5)
        if aval["FLAGS"]['extlen']:
            flags = flags | (1<<4)

        rflgs = struct.pack('BB', flags & 0xff, attr & 0xff)
        if   attr == bgp.PATH_ATTRIBUTES["ORIGIN"]:
            rval = struct.pack('B', aval["V"])
                        
        elif attr == bgp.PATH_ATTRIBUTES["AS_PATH"]:
            if aval["V"] != None:
                for pseg in aval["V"]:
                    tl = struct.pack('BB', pseg["T"], pseg["L"])
                    val = ""
                    for v in pseg["V"]:
                        val = struct.pack('>%dsH' % len(val), val, v)
                rval = struct.pack('%ds%ds' % (len(tl), len(val)), tl, val)
            else:
                rval = struct.pack('BB',
                                   bgp.AS_PATH_SEG_TYPES['SEQUENCE'], 0)

        elif attr == bgp.PATH_ATTRIBUTES["NEXT_HOP"]:
            rval = struct.pack('>L', aval["V"])

        elif attr == bgp.PATH_ATTRIBUTES["MULTI_EXIT_DISCRIMINATOR"]:
            rval = struct.pack('>L', aval["V"])
            
        elif attr == bgp.PATH_ATTRIBUTES["LOC_PREF"]:
            rval = struct.pack('>L', aval["V"])

        elif attr == bgp.PATH_ATTRIBUTES["ATOMIC_AGGR"]:
            rval = ""

        elif attr == bgp.PATH_ATTRIBUTES["AGGREGATOR"]:
            rval = struct.pack(">HL", aval["V"][0], aval["V"][1])

        elif attr == bgp.PATH_ATTRIBUTES["COMMUNITY"]:
            rval = ""
            for v in aval["V"]:
                rval = struct.pack('>%ds4s' % len(rval), rval, v)

        elif attr == bgp.PATH_ATTRIBUTES["ORIGINATOR_ID"]:
            rval = struct.pack('>L', aval["V"])

        elif attr == bgp.PATH_ATTRIBUTES["CLUSTER_LIST"]:
            rval = ""
            for v in aval["V"]:
                rval = struct.pack('>%dsL' % len(rval), rval, v)

        else:
            print '[ *** Unsupported attribute:', \
                  bgp.PATH_ATTRIBUTES[attr], '*** ]'
            return None

        if aval["FLAGS"]['extlen']:            
            rlen = struct.pack('>H', len(rval) & 0xffff)
        else:
            rlen = struct.pack('>B', len(rval) & 0xff)

        return struct.pack('%ds%ds%ds' % (len(rflgs), len(rlen), len(rval)),
                           rflgs, rlen, rval)

    msg_tm = rv["H"]["TIME"]
    src_as, src_ip = rv["H"]["SRC_AS"], rv["H"]["SRC_IP"]
    ifc, afi = rv["H"]["IFC"], rv["H"]["AFI"]

    for pfx in rv["V"]["V"]["UNFEASIBLE"]:
        del TABLE[pfx]

    astr = ""
    for attr in rv["V"]["V"]["PATH_ATTRS"].keys():
        rstr = mkStr(attr, rv["V"]["V"]["PATH_ATTRS"][attr])
        astr = struct.pack('%ds%ds' % (len(astr), len(rstr)),
                           astr, rstr)

        
    for pfx in rv["V"]["V"]["FEASIBLE"]:
        TABLE[pfx] = {"TIME"   : msg_tm,
                      "PEER_IP": src_ip,
                      "PEER_AS": src_as,
                      "ATTRS"  : astr,
                      }

#-------------------------------------------------------------------------------

def dumpTable():

    now = time.time()
    date_str = time.strftime(".%Y-%m-%d_%H.%M.%S", time.gmtime(LAST_TM))
    of = open(OUTPUT_F + date_str, 'w+b')

    error('dumping...')

    seq_no = 0
    for (pfx, plen) in TABLE.keys():
        attr_len = len(TABLE[(pfx,plen)]['ATTRS'])
        common_hdr = struct.pack('>HH', VIEW_NO, seq_no)
        entry      = struct.pack('>4sBBLLHH%ds' % attr_len,
                                 pfx, plen, STATUS,
                                 TABLE[(pfx,plen)]['TIME'],
                                 TABLE[(pfx,plen)]['PEER_IP'],
                                 TABLE[(pfx,plen)]['PEER_AS'],
                                 attr_len, TABLE[(pfx,plen)]['ATTRS'])
        
        mrt_hdr = struct.pack('>LHHL',
                              now,
                              mrtd.MSG_TYPES['TABLE_DUMP'],
                              mrtd.TABLE_DUMP_SUBTYPES['IP'],
                              attr_len+\
                              mrtd.TABLE_DUMP_HDR_LEN+\
                              bgp.TABLE_DUMP_ENTRY_HDR_LEN)

        of.write('%s%s%s' % (mrt_hdr, common_hdr, entry))
        seq_no = seq_no + 1

    error('[%d] entries...' % len(TABLE.keys()) )

    of.close()

################################################################################

if __name__ == "__main__":

    VERBOSE  = 1
    START_T  = -1
    INTERVAL = -1
    LAST_TM  = -1

    OUTPUT_F = 'bview'
    TABLE_F = None
    TABLE   = {}
    
    #---------------------------------------------------------------------------

    def usage():

        print """Usage: %s [ options ] <filenames>:
        -h|--help       : Help
        -q|--quiet      : Be quiet
        -v|--verbose    : Be verbose

        -f|--file       : Filename prefix for output
        -s|--start-time : Start time of packets of interest [inclusive]
        -i|--interval   : Table dump invterval (minutes)
        -t|--table      : Initial table [def.: none]""" %\
            (os.path.basename(sys.argv[0]),)
        sys.exit(0)
    
    #---------------------------------------------------------------------------

    if len(sys.argv) < 2:
        usage()
        
    try:
        opts, args =\
              getopt.getopt(sys.argv[1:],
                            "hqvf:s:i:t:",
                            ("help", "quiet", "verbose",
                             "file=", "start-time=", "interval=", "table=" ))
    except (getopt.error):
        usage()

    for (x, y) in opts:        
        if x in ('-h', '--help'):
            usage()

        elif x in ('-q', '--quiet'):
            VERBOSE = 0

        elif x in ('-v', '--verbose'):
            VERBOSE = 2

        elif x in ('-f', '--file'):
            OUTPUT_F = y

        elif x in ('-s', '--start-time'):
            START_T = time.mktime(time.strptime(y))

        elif x in ('-i', '--interval'):
            INTERVAL = string.atol(y) * MINS_TO_SECS

        elif x in ('-t', '--table'):
            TABLE_F = y

        else:
            usage()

    filenames = args
    if not filenames:
        usage()

    #---------------------------------------------------------------------------

    NEXT_DUMP = START_T + INTERVAL

    # seed from dump if required here

    if TABLE_F:
        cnt = 0
        error('[ %s ] initializing table...' % TABLE_F)
        try:
            mrt = mrtd.Mrtd(TABLE_F, "rb")
            while 1:
                rv = mrt.parse(mrt.read(), VERBOSE)
                cnt = cnt + 1
                if rv["T"] == mrtd.MSG_TYPES["TABLE_DUMP"]:
                    for v in rv["V"]:
                        pfx = v["V"]["PREFIX"]
                        entry = { "TIME"   : v["V"]["UPTIME"],
                                  "PEER_IP": v["V"]["PEER_IP"],
                                  "PEER_AS": v["V"]["PEER_AS"],
                                  "ATTRS"  : v["V"]["ATTRS"],
                                  }

                        TABLE[pfx] = entry
        except (mrtd.EOFExc):
            error("end of file: %u messages\n" % cnt)
        except (KeyboardInterrupt):
            error("interrupted: %u messages\n" % cnt)
        mrt.close()
        error('done\n')

    print 'init entries:', `len(TABLE.keys())`

    # process UPDATE files
    
    for fn in filenames:
        cnt = 0
        try:
            error('[ %s ] parsing...' % fn)
            mrt = mrtd.Mrtd(fn, "rb")
            while 1:
                msg = mrt.read()
                cnt = cnt + 1
                if (START_T < 0) or (msg[0] >= START_T):
                    rv = mrt.parse(msg, VERBOSE)
                    if ((rv["T"] == mrtd.MSG_TYPES["PROTOCOL_BGP"] and
                         rv["ST"] == mrtd.BGP_SUBTYPES["UPDATE"])
                        or
                        (rv["T"] == mrtd.MSG_TYPES["PROTOCOL_BGP4MP"] and
                         rv["ST"] == mrtd.BGP4MP_SUBTYPES["MESSAGE"] and
                         rv["V"]["T"] == bgp.MSG_TYPES["UPDATE"])
                        or
                        (rv["T"] == mrtd.MSG_TYPES["PROTOCOL_BGP4PY"] and
                         rv["ST"] == mrtd.BGP4MP_SUBTYPES["MESSAGE"] and
                         rv["V"]["T"] == bgp.MSG_TYPES["UPDATE"])
                        ):

                        processEntry(rv)

                        LAST_TM = msg[0]
                        if (LAST_TM > NEXT_DUMP and NEXT_DUMP > START_T):
                            dumpTable()
                            NEXT_DUMP = NEXT_DUMP + INTERVAL

        except (mrtd.EOFExc):
            error("end of file: %u messages..." % cnt)
        except (KeyboardInterrupt):
            error("interrupted: %u messages..." % cnt)
        error('done\n')
        mrt.close()

    # do a final table dump

    dumpTable()
    
    sys.exit(0)

################################################################################
################################################################################
